### PR TITLE
[FEAT] Background Location Tracking Task

### PR DESCRIPTION
**Pull request by dhruv423**
## Overview

Implements the background tracking task once a user turns on location (aka TabNavigator). Background Location stops if the user closes the app (quit app) or logs out. It will start up again after the user opens the app.

This change is a

- [ ] Bug fix
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that will change existing functionality)
- [ ] Documentation update

## Description

 - Start Background Task when TabNavigator is loaded
 - Stop Background Task when user logs out
 - Need to add API call for location update
 - https://forums.expo.dev/t/how-to-setstate-from-within-taskmanager/26630/5, we can pass the updated location to a service file that holds the info (or we can do a pub-sub approach like the link or we use Redux) and we can access that in other pages and use it for necessary API calls.

## Motivation/Links

To get users location

### CODE REVIEW

File: src/hooks/backgroundTrackingTask.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import * as TaskManager from "expo-task-manager";
+
+export const LOCATION_TASK_NAME = "LOCATION_BACKGROUND_TRACKING";
+
+// Define the background task for location tracking
+TaskManager.defineTask(
+  LOCATION_TASK_NAME,
+  async ({ data, error }: { data: any; error: any }) => {
+    if (error) {
+      console.error(error);
```
Comment (dhruv423) [inline src/hooks/backgroundTrackingTask.tsx:13]: Not sure what we want to do here, if it errors

**Comment by benjaminkostiuk** [comment]
> Need to add API call for location update
> we can pass the updated location...

Are these things to add to this PR? Or future ones?

**Review by benjaminkostiuk — APPROVED** [review]
Looks good! Not sure if this is a stepping stone PR or if you want to implement more stuff in this one before merging.

File: src/hooks/backgroundTrackingTask.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      return;
+    }
+    if (data) {
+      // Call API to update user location
```
[the remark this answers is not in the document]
Comment (dhruv423) [reply src/hooks/backgroundTrackingTask.tsx:17]: Oh I didn't realize we have the API ready, I will hook it up in this PR. But yeah any ideas for the user token?

File: src/contexts/auth.tsx [hunk 3]
```diff
@@ -39,6 +41,12 @@ const signIn = async () => {
 const signOut = async () => {
   await GoogleSignin.revokeAccess(); // Allows selection of Google account after sign-out
   await auth().signOut();
+  const hasStarted = await Location.hasStartedLocationUpdatesAsync(
+    LOCATION_TASK_NAME
+  );
+  if (hasStarted) {
+    await Location.stopLocationUpdatesAsync(LOCATION_TASK_NAME);
```
[the remark this answers is not in the document]
Comment (dhruv423) [reply src/contexts/auth.tsx:48]: Don't think it fails but I can add it

File: src/hooks/backgroundTrackingTask.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (benjaminkostiuk) [reply src/hooks/backgroundTrackingTask.tsx:17]: I think the easiest way is just to turn this into a function that takes in a user and call it from a component that has access to the user. Probably tab navigator.

File: src/contexts/auth.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (benjaminkostiuk) [reply src/contexts/auth.tsx:48]: Ahh, just remembered that I always handle the errors where I call this. So no need.

File: src/hooks/backgroundTrackingTask.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    if (data) {
+      const { latitude, longitude } = data.locations[0].coords;
+      // Delaying the update of the geolocation state to avoid getting setRecoil is not a function error
+      setTimeout(() => {
+        setRecoil(userGeolocationState, {
+          latitude,
+          longitude,
+          timestamp: Date.now(),
+        });
+      }, 1000);
```
Comment (benjaminkostiuk) [inline src/hooks/backgroundTrackingTask.tsx:29]: So this is delaying it by 1 second *each time* we get an update for the location. Instead you want to wrap this whole piece of code to delay it until it's loaded, not delaying the update each time.

**Review by benjaminkostiuk — APPROVED** [review]
Small fix for the setTimeout, otherwise we're golden. Thanks for working through the bugs for this.

**Pull request merged**